Apache Pulsar 2.9.0 release notes

eolivelli opened the pull request.

Anonymitaet commented:
@eolivelli thanks for your contribution!

Can you keep consistent with other release notes? Thanks

This PR
![image](https://user-images.githubusercontent.com/50226895/138016997-80980eb6-132d-464b-b36a-27dda84cd889.png)

Other release notes:
![image](https://user-images.githubusercontent.com/50226895/138019730-2f1cd000-e921-4ccb-8bb4-34360add379f.png)

Anonymitaet commented:
Please keep consistent for cases (normally, we use sentence cases) and formats (you can use `PIP XX` or `PIP-XX`, no matter which one you use, just keep consistent)

I do not highlight every place to be modified, can you please check all occurrences? thanks!

![image](https://user-images.githubusercontent.com/50226895/138025349-c37b688e-59e1-49ba-ab3e-44b4b4f0622d.png)

michaeljmarshall commented on site2/website/release-notes.md:
```diff
@@ -1,6 +1,439 @@
+- Avoid to infinitely split bundle. [11937](https://github.com/apache/pulsar/pull/11937)
+- Update the default log pattern. [11935](https://github.com/apache/pulsar/pull/11935)
+- [Transaction] add method to clear up transaction buffer snapshot. [11934](https://github.com/apache/pulsar/pull/11934)
+- [pulsar-client]Add a optional params scope for pulsar oauth2 client. [11931](https://github.com/apache/pulsar/pull/11931)
```
Nit: this was an update to the Java client.

eolivelli replied:
yes. it makes sense to note it in the section above, thanks

eolivelli commented on site2/website/release-notes.md:
```diff
@@ -1,6 +1,439 @@
+### News and noteworthy
+- PIP-45 Pluggable metadata interface introduced many changes about ZooKeeper Metadata Management: consistency, resilience, stability, tech debt reduction (less code duplication)
+- Pulsar IO: Oracle Debezium Connector, new Schema Aware Elasticsearch Sink Connector
+- Many improvements to the Pulsar clients, with PIP-83, PIP-91, PIP-96
```
there is a list below

eolivelli commented on site2/website/release-notes.md:
```diff
@@ -1,6 +1,439 @@
+- [Python Schema] Support setting namespace for python schema. [12175](https://github.com/apache/pulsar/pull/12175)
+- [C++] Support setting priority for the consumer. [12076](https://github.com/apache/pulsar/pull/12076)
+- [C++] Fixed attempting to connect to multiple IP addresses. [11958](https://github.com/apache/pulsar/pull/11958)
+- [Python] Expose Client.shutdown() method. [11955](https://github.com/apache/pulsar/pull/11955)
```
because Python is basically a wrapper over the C++ client, I will change the title

Anonymitaet replied:
This one is not resolved yet

eolivelli commented on site2/website/release-notes.md:
```diff
@@ -1,6 +1,439 @@
+- Fix the potential race condition in the BlobStore readhandler. [12123](https://github.com/apache/pulsar/pull/12123)
+
+
+#### Misc
```
There are hundreds of entries.
I just move above the most interesting things, it is pretty useless for the users.

Probably we can drop this list at all.
Release notes should be brief and give a good overview of the things that you should know while upgrading.
The full list is already on github.

I am not sure it is worth to spend 2-3 hours in doing a categorisation

Anonymitaet replied:
Agree that we do not include them in the release notes if they are minor fixes/edits.

eolivelli commented:
@Anonymitaet I did another pass. PTAL

I will start discussion on dev@ about how we create these release notes.

The way we are doing is not good:
- there is too much manual work to be done
- we are giving too mach useless information to the users (the list of PRs is useless, advanced users have "git log" and the Git Hub UI, non advanced users really don't care)

We should focus on the "News and Noteworthy" and drop the boilerplate, also this page is becoming bigger and bigger

I would commit this patch as it is now and move forward with the release.
What do you think ?

Anonymitaet commented:
Hi @eolivelli this PR:
- involves many changes beyond docs
- there is no `.md` file (no content) for release notes 

is this an intended behavior?

![image](https://user-images.githubusercontent.com/50226895/143982739-8d4e5700-4d97-418b-bb5a-ce8fbefc829a.png)

eolivelli commented:
@Anonymitaet GitHub went crazy when I merged with current master.
I have fixed this PR
Sorry about that

Anonymitaet commented:
> @Anonymitaet I did another pass. PTAL
> 
> I will start discussion on dev@ about how we create these release notes.
> 
> The way we are doing is not good:
> 
> * there is too much manual work to be done
> * we are giving too mach useless information to the users (the list of PRs is useless, advanced users have "git log" and the Git Hub UI, non advanced users really don't care)
> 
> We should focus on the "News and Noteworthy" and drop the boilerplate, also this page is becoming bigger and bigger
> 
> I would commit this patch as it is now and move forward with the release. What do you think ?

+1, just highlight the important items and summarize the issue and solutions, like what we did for previous releases:

Example:
2.8.1 https://pulsar.apache.org/blog/2021/09/23/Apache-Pulsar-2-8-1
![image](https://user-images.githubusercontent.com/50226895/144009748-6dc9d1c0-ebcc-4467-aded-62ebcfe8d0c3.png)

Anonymitaet commented:
![image](https://user-images.githubusercontent.com/50226895/144010229-b1c64992-37fa-4285-b77a-c248bb162124.png)

Anonymitaet commented:
![image](https://user-images.githubusercontent.com/50226895/144010553-3cde3f42-613b-4e15-8de7-97ab6933794c.png)

Anonymitaet commented:
> Please keep consistent for cases (normally, we use sentence cases) and formats (you can use `PIP XX` or `PIP-XX`, no matter which one you use, just keep consistent)
> 
> I do not highlight every place to be modified, can you please check all occurrences? thanks!
> 
> ![image](https://user-images.githubusercontent.com/50226895/138025349-c37b688e-59e1-49ba-ab3e-44b4b4f0622d.png)

this one is not resolved yet

Anonymitaet commented:
> If you want to add a short summary (`[Functions]`, `[Pulsar IO Kinesis]`, `[xx]`) for items, can you add the summary for all items and keep consistent?
> 
> Same comments for other headings like `Java client`, `C++ client`, and so on
> 
> ![image](https://user-images.githubusercontent.com/50226895/138025653-1486b6a8-e002-4d12-9920-257d8e8315b8.png)

this one is not resolved yet

Anonymitaet reviewed (commented):
Can you keep tense consistent? For example, use `past tense` for verbs since engineers finished them before.

Example: https://cloud.google.com/release-notes#November_29_2021

![image](https://user-images.githubusercontent.com/50226895/144011879-4c52eb73-1258-4c4e-b129-9ae8e574e8c5.png)

Anonymitaet commented on site2/website/release-notes.md:
```diff
@@ -1,6 +1,438 @@
+- Geo-replication improvements: PIP-88 replicate schemas across clusters
+- Apache Kafka connect sinks can be run as Pulsar sinks [#9927](https://github.com/apache/pulsar/pull/9927)
+
+### Breaking changes
```
What's the difference between `News and noteworthy` and `Breaking changes`? Can we put them into one section?

eolivelli replied:
This is very common, I saw this in many other projects and products:

- `News and noteworth`: cool stuff, only good and nothing bad
- `Breaking changes`: this is usually painful for people who upgrade

So it is better to keep these sections separated

Anonymitaet commented on site2/website/release-notes.md:
```diff
@@ -1,6 +1,438 @@
+- PIP-85: [pulsar-io] pass pulsar client via context to connector. [11056](https://github.com/apache/pulsar/pull/11056)
+
+### Java Client
+- [Java Client] Use failPendingMessages to ensure proper cleanup. [12259](https://github.com/apache/pulsar/pull/12259)
+- [Java Client] Remove unnecessary stats incremement. [12229](https://github.com/apache/pulsar/pull/12229)
+- [Java Client] Remove unnecessary synchronization and simplify initialization. [12228](https://github.com/apache/pulsar/pull/12228)
+- [pulsar-client] Add partition-change API for producer or consumer interceptors. [12287](https://github.com/apache/pulsar/pull/12287)
+- Fix deadLetterPolicy is not working with key shared subscription under partitioned topic. [12148](https://github.com/apache/pulsar/pull/12148)
+- [Client] Fix endless receiveAsync loop in MultiTopicsConsumer. [12044](https://github.com/apache/pulsar/pull/12044)
+- [Client] Reduce redundant FLOW requests for non-durable multi-topics consumer. [11802](https://github.com/apache/pulsar/pull/11802)
+- [pulsar-client] Set and return topic names on message API. [11743](https://github.com/apache/pulsar/pull/11743)
+- PIP-91: Separate lookup timeout from operation timeout. [11627](https://github.com/apache/pulsar/pull/11627)
+- [Java Client] Switch from pretty print to compact print for configs. [11609](https://github.com/apache/pulsar/pull/11609)
+- Java Client: remove usage of reflection while using Pulsar Implementation classes. [11636](https://github.com/apache/pulsar/pull/11636)
+- [pulsar-client]Add a optional params scope for pulsar oauth2 client. [11931](https://github.com/apache/pulsar/pull/11931)
```
1. Since these items belong to `Java client`, can we delete [Java client] for each item to make content more concise?
![image](https://user-images.githubusercontent.com/50226895/144012351-79de642d-d1dc-4192-a8c9-2f3c9cd3de90.png)

2. Same comments for other sections

eolivelli replied:
done

Anonymitaet commented on site2/website/release-notes.md:
```diff
@@ -1,6 +1,438 @@
+- Java Client: remove usage of reflection while using Pulsar Implementation classes. [11636](https://github.com/apache/pulsar/pull/11636)
+- [pulsar-client]Add a optional params scope for pulsar oauth2 client. [11931](https://github.com/apache/pulsar/pull/11931)
+
+### C++ and Python Client
```
Can we separate this big part into `### C++ client` and `### Python client` to make content more clear?

eolivelli replied:
yes. done

Anonymitaet commented:
> @Anonymitaet I did another pass. PTAL
> 
> I will start discussion on dev@ about how we create these release notes.
> 
> The way we are doing is not good:
> 
> * there is too much manual work to be done
> * we are giving too mach useless information to the users (the list of PRs is useless, advanced users have "git log" and the Git Hub UI, non advanced users really don't care)
> 
> We should focus on the "News and Noteworthy" and drop the boilerplate, also this page is becoming bigger and bigger
> 
> I would commit this patch as it is now and move forward with the release. What do you think ?

+1. We can include only highlights rather explain everything in the release notes

eolivelli reviewed (commented):
@Anonymitaet 
I have addressed your comments.

We definitely have to find a better way to create these release notes for the future :-)

eolivelli commented:
> > Please keep consistent for cases (normally, we use sentence cases) and formats (you can use `PIP XX` or `PIP-XX`, no matter which one you use, just keep consistent)
> > I do not highlight every place to be modified, can you please check all occurrences? thanks!
> > ![image](https://user-images.githubusercontent.com/50226895/138025349-c37b688e-59e1-49ba-ab3e-44b4b4f0622d.png)
> 
> this one is not resolved yet

I have removed "PIP-89"


This is really a list of "Commit messages".
I am not sure it is worth to change them so much.

The spirit of this page **currently** is to report the PRs, if change the titles to much there will be no good match.

I believe that we should create a new way of creating the release notes

eolivelli commented:
> Can you keep tense consistent? For example, use `past tense` for verbs since engineers finished them before.

As said, from my point of view it is better to not change that much the commit title of the PRs

Anonymitaet commented:
> > Can you keep tense consistent? For example, use `past tense` for verbs since engineers finished them before.
> 
> As said, from my point of view it is better to not change that much the commit title of the PRs

As least, tense should be kept consistent in the release notes, whether "present tense" or "past tense", so can you check that?

Anonymitaet commented:
> > > Please keep consistent for cases (normally, we use sentence cases) and formats (you can use `PIP XX` or `PIP-XX`, no matter which one you use, just keep consistent)
> > > I do not highlight every place to be modified, can you please check all occurrences? thanks!
> > > ![image](https://user-images.githubusercontent.com/50226895/138025349-c37b688e-59e1-49ba-ab3e-44b4b4f0622d.png)
> > 
> > 
> > this one is not resolved yet
> 
> I have removed "PIP-89"
> 
> This is really a list of "Commit messages". I am not sure it is worth to change them so much.
> 
> The spirit of this page **currently** is to report the PRs, if change the titles to much there will be no good match.
> 
> I believe that we should create a new way of creating the release notes

A good release note is a clear and concise summary of changes rather than a copy of all commit messages, which means each PR and its description should be checked carefully. If a PR title is not good enough, in the release note, we can re-write the PR title to make it more accurate and readable. No need to worry about the "different titles" since the PR link is added at the end. Users can know the details if they want.

Besides, the consistent format is the basic requirement for a release note.
I check this several times, but find that the following contents are still in inconsistent formats:
- Cases (upper case, lower case)
- Links, eg. PIP XX, PIP-XX, `[#XXXX](xxx)`, `[XXXX](xxx)`
Below are just some examples, can you check them all? 
![image](https://user-images.githubusercontent.com/50226895/144387360-1410514f-8f02-43ff-badc-fc66fe1c231c.png)
![image](https://user-images.githubusercontent.com/50226895/144387716-e5892824-b55f-418f-8ab9-17cd324a1455.png)

Anonymitaet commented on site2/website/release-notes.md:
```diff
@@ -1,6 +1,440 @@
+### Tiered Storage
+- Fix the potential race condition in the BlobStore readhandler. [12123](https://github.com/apache/pulsar/pull/12123)
+
+### Misc
```
I remember that you've made a comment about `Misc`. If this part is not categorized into several small sections, the readability is not good since users do not have enough patience to wade through a large chunk of information.

eolivelli replied:
I believe that this part is pretty useless for users and we could remove it at all.

Having such detailed list of PRs is not useful for anyone.
Advanced users use "git log" or they can always use the GH interface to run thru the list of PRs.

I am sorry but I don't want to spend hours in classifying 2-3 hundreds of PRs

Anonymitaet replied:
Retaining some important PRs in `Misc` might be a good approach. 
Besides, it is better to invite other people to review this PR and give their opinions.

BewareMyPower replied:
I agree that `Misc` part is useless for users and we should remove it. See https://pulsar.apache.org/en/release-notes/#apache-pulsar-release-notes for example, the release notes of 2.8.1 and 2.8.0 don't contain this part.

But I also found many PRs in this part should be noted in other parts. For example,

```
- fix publish_time not set error when broker entry metadata enable without AppendBrokerTimestampMetadataInterceptor. [11014](https://github.com/apache/pulsar/pull/11014)
```

fixes a very common issue of Pulsar 2.8.0 when enabling KoP.

> I am sorry but I don't want to spend hours in classifying 2-3 hundreds of PRs

It might took some time but I think it's necessary. From my perspective, the release notes of this PR looks very terrible, especially compared with the existing release notes for 2.8.0 and 2.8.1.

Some semantics errors or inconsistency are acceptable. Take release notes of 2.8.1 as example,

```
Add metrics for writing or reading size of cursor #11500
Added missing configuration entries #11095
```

It's totally acceptable IMO.

But I think the PR description (in this PR) like

```
- fix lost message issues 12221. [12223](https://github.com/apache/pulsar/pull/12223)
```

should be fixed. No one knows what's the issue of 12221 only from the website.

The readers of release notes are usually those who want to check if this release has fixed some specific bugs or supported some features. The classification is very important though it could take much time.

BewareMyPower replied:
The `fix lost message issues 12221` title is a bad work done by who merged this PR. But we cannot avoid this problem since we have many committers now. I think release manager is responsible to do the cleanup for it.

Maybe we can add some enhancement for the process of merging PRs in future, but currently we should do it manually.

eolivelli commented:
@Anonymitaet I have removed the misc section, moving more PRs to other sections.
I have fixed all the titles of the PRs

I believe that this patch is in good shape and we can merge it.
I would like to move forward with the release process of 2.9.0.

There are many other steps to perform and we should start packaging 2.9.1 soon

please take another look

michaeljmarshall commented on site2/website/release-notes.md:
```diff
@@ -1,8 +1,107 @@
-### 2.8.1
+### 2.9.0
+#### 2021-11-25
+### Broker
```
```suggestion
```

michaeljmarshall commented on site2/website/release-notes.md:
```diff
@@ -1,8 +1,107 @@
+- Support protobuf schema for pulsar function. [11709](https://github.com/apache/pulsar/pull/11709)
+- Support KEY_BASED batch builder for Java based functions and sources. [11706](https://github.com/apache/pulsar/pull/11706)
+- Upgrade Go client version to 0.6.0. [11477](https://github.com/apache/pulsar/pull/11477)
+- Stop calling the deprecated method Thread.stop() when stopping the function thread in ThreadRuntim. [11401](https://github.com/apache/pulsar/pull/11401)
```
```suggestion
- Stop calling the deprecated method Thread.stop() when stopping the function thread in ThreadRuntime. [11401](https://github.com/apache/pulsar/pull/11401)
```

michaeljmarshall commented on site2/website/release-notes.md:
```diff
@@ -1,8 +1,107 @@
+- Broker auto refresh bk-client certs to avoid cnx failure after cert refresh. [12107](https://github.com/apache/pulsar/pull/12107)
+- Add multi roles support for authorization. [11341](https://github.com/apache/pulsar/pull/11341)
+
+# Docker
```
```suggestion
### Docker
```

michaeljmarshall commented:
Merging now. If anyone would like to adjust the notes, please submit a PR.
